pkg/segment/segexecution.go

```diff
@@ -157,52 +157,99 @@ func HelperQueryArithmeticAndLogical(queryOps []structs.QueryArithmetic, resMap
-					valueRHS := resultRHS.Results[groupID][timestamp]
+					valueRHS := resultRHS.Results[rGroupID][timestamp]
 					switch queryOp.Operation {
 					case utils.LetAdd:
-						finalResult[groupID][timestamp] = valueLHS + valueRHS
+						finalResult[lGroupID][timestamp] = valueLHS + valueRHS
 					case utils.LetDivide:
 						if valueRHS == 0 {
 							continue
 						}
-						finalResult[groupID][timestamp] = valueLHS / valueRHS
+						finalResult[lGroupID][timestamp] = valueLHS / valueRHS
 					case utils.LetMultiply:
-						finalResult[groupID][timestamp] = valueLHS * valueRHS
+						finalResult[lGroupID][timestamp] = valueLHS * valueRHS
 					case utils.LetSubtract:
-						finalResult[groupID][timestamp] = valueLHS - valueRHS
+						finalResult[lGroupID][timestamp] = valueLHS - valueRHS
 					case utils.LetGreaterThan:
 						if valueLHS > valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
 					case utils.LetGreaterThanOrEqualTo:
 						if valueLHS >= valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
 					case utils.LetLessThan:
 						if valueLHS < valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
 					case utils.LetLessThanOrEqualTo:
 						if valueLHS <= valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
 					case utils.LetEquals:
 						if valueLHS == valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
 					case utils.LetNotEquals:
 						if valueLHS != valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
+					case utils.LetAnd:
+						finalResult[lGroupID][timestamp] = valueLHS
+					case utils.LetOr:
+						finalResult[lGroupID][timestamp] = valueLHS
+					case utils.LetUnless:
+						finalResult[lGroupID][timestamp] = valueLHS
+					}
+				}
+			}
+			if queryOp.Operation == utils.LetOr || queryOp.Operation == utils.LetUnless {
+				for rGroupID, tsRHS := range resultRHS.Results {
+					labelStr := ""
+					if len(rGroupID) >= len(resultLHS.MetricName) {
+						labelStr = rGroupID[len(resultLHS.MetricName):]
+					}
+
+					// For unless op, all matching elements in both vectors are dropped
+					if queryOp.Operation == utils.LetUnless {
+						lGroupID := resultLHS.MetricName + labelStr
+						delete(finalResult, lGroupID)
+						continue
+					}
+
+					_, exists := labelStrSet[labelStr]
+					// If exists, which means we already add that label set when traversing the resultLHS
+					if exists {
+						continue
+					}
+
+					finalResult[rGroupID] = make(map[uint32]float64)
+					for timestamp, valueRHS := range tsRHS {
+						finalResult[rGroupID][timestamp] = valueRHS
 					}
 				}
 			}
```

Comment:
I think handling `LetOr` and `LetUnless` as two different `if`s will make the code more clear. Especially if we have to add more special cases later on, handling them all in one `if` is going to get confusing